In [1]:
import pandas as pd 
import numpy as np 

In [43]:
order_item = pd.read_csv(r"C:\works\learnings\story_telling_insight_analysis\Data_folder\olist_order_items_dataset.csv")
order_review = pd.read_csv(r"C:\works\learnings\story_telling_insight_analysis\Data_folder\olist_order_reviews_dataset.csv")
order_dataset = pd.read_csv(r"C:\works\learnings\story_telling_insight_analysis\Data_folder\olist_orders_dataset.csv")
product_details = pd.read_csv(r"C:\works\learnings\story_telling_insight_analysis\Data_folder\olist_products_dataset.csv")
seller_details  = pd.read_csv(r"C:\works\learnings\story_telling_insight_analysis\Data_folder\olist_sellers_dataset.csv")
order_payment = pd.read_csv(r"C:\works\learnings\story_telling_insight_analysis\Data_folder\olist_order_payments_dataset.csv")
geo_location = pd.read_csv(r"C:\works\learnings\story_telling_insight_analysis\Data_folder\olist_geolocation_dataset.csv")
customer_data = pd.read_csv(r"C:\works\learnings\story_telling_insight_analysis\Data_folder\olist_customers_dataset.csv")

In [44]:
customer_data.head()

,customer_id,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,06b8999e2fba1a1fbc88172c00ba8bc7,861eff4711a542e4b93843c6dd7febb0,14409,franca,SP
1,18955e83d337fd6b2def6b18a428ac77,290c77bc529b7ac935b93aa66c333dc3,9790,sao bernardo do campo,SP
2,4e7b3e00288586ebd08712fdd0374a03,060e732b5b29e8181a18229c7b0b2b5e,1151,sao paulo,SP
3,b2b6027bc5c5109e529d4dc6358b12c3,259dac757896d24d7702b9acbbff3f3c,8775,mogi das cruzes,SP
4,4f2d8ab171c80ec8364f7c12e35b23ad,345ecd01c38d18a9036ed96c73b8d066,13056,campinas,SP


In [3]:
order_dataset.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18 00:00:00
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13 00:00:00
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04 00:00:00
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15 00:00:00
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26 00:00:00


In [4]:
order_item.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14


In [5]:
order_review.head()

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,7bc2406110b926393aa56f80a40eba40,73fc7af87114b39712e6da79b0a377eb,4,NaN,NaN,2018-01-18 00:00:00,2018-01-18 21:46:59
1,80e641a11e56f04c1ad469d5645fdfde,a548910a1c6147796b98fdf73dbeba33,5,NaN,NaN,2018-03-10 00:00:00,2018-03-11 03:05:13
2,228ce5500dc1d8e020d8d1322874b6f0,f9e4b658b201a9f2ecdecbb34bed034b,5,NaN,NaN,2018-02-17 00:00:00,2018-02-18 14:36:24
3,e64fb393e7b32834bb789ff8bb30750e,658677c97b385a9be170737859d3511b,5,NaN,Recebi bem antes do prazo estipulado.,2017-04-21 00:00:00,2017-04-21 22:02:06
4,f7c4243c7fe1938f181bec41a392bdeb,8e6bfb81e283fa7e4f11123a3fb894f1,5,NaN,Parabéns lojas lannister adorei comprar pela I...,2018-03-01 00:00:00,2018-03-02 10:26:53


In [6]:
product_details.head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
0,1e9e8ef04dbcff4541ed26657ea517e5,perfumaria,40.0,287.0,1.0,225.0,16.0,10.0,14.0
1,3aa071139cb16b67ca9e5dea641aaa2f,artes,44.0,276.0,1.0,1000.0,30.0,18.0,20.0
2,96bd76ec8810374ed1b65e291975717f,esporte_lazer,46.0,250.0,1.0,154.0,18.0,9.0,15.0
3,cef67bcfe19066a932b7673e239eb23d,bebes,27.0,261.0,1.0,371.0,26.0,4.0,26.0
4,9dc1a7de274444849c219cff195d0b71,utilidades_domesticas,37.0,402.0,4.0,625.0,20.0,17.0,13.0


In [45]:
geo_location.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1037,-23.545621,-46.639292,sao paulo,SP
1,1046,-23.546081,-46.644820,sao paulo,SP
2,1046,-23.546129,-46.642951,sao paulo,SP
3,1041,-23.544392,-46.639499,sao paulo,SP
4,1035,-23.541578,-46.641607,sao paulo,SP


In [7]:
seller_details.head()

,seller_id,seller_zip_code_prefix,seller_city,seller_state
0,3442f8959a84dea7ee197c632cb2df15,13023,campinas,SP
1,d1b65fc7debc3361ea86b5f14c68d2e2,13844,mogi guacu,SP
2,ce3ad9de960102d0677a81f5d0bb7b2d,20031,rio de janeiro,RJ
3,c0f3eea2e14555b6faeea3dd58c1b1c3,4195,sao paulo,SP
4,51a04a8a6bdcb23deccc82b0b80742cf,12914,braganca paulista,SP


In [8]:
order_payment.head()

,order_id,payment_sequential,payment_type,payment_installments,payment_value
0,b81ef226f3fe1789b1e8b2acac839d17,1,credit_card,8,99.33
1,a9810da82917af2d9aefd1278f1dcfa0,1,credit_card,1,24.39
2,25e8ea4e93396b6fa0d3dd708e76c1bd,1,credit_card,1,65.71
3,ba78997921bbcdc1373bb41e913ab953,1,credit_card,8,107.78
4,42fdf880ba16b47b59251dd489d4441a,1,credit_card,2,128.45


In [9]:
order_dataset["order_delivered_customer_date"] = pd.to_datetime(order_dataset["order_delivered_customer_date"])
order_dataset["order_estimated_delivery_date"] = pd.to_datetime(order_dataset["order_estimated_delivery_date"])

In [10]:
order_dataset["Delay"] = order_dataset["order_delivered_customer_date"] - order_dataset["order_estimated_delivery_date"]
order_dataset.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,Delay
0,e481f51cbdc54678b7cc49136f2d6af7,9ef432eb6251297304e76186b10a928d,delivered,2017-10-02 10:56:33,2017-10-02 11:07:15,2017-10-04 19:55:00,2017-10-10 21:25:13,2017-10-18,-8 days +21:25:13
1,53cdb2fc8bc7dce0b6741e2150273451,b0830fb4747a6c6d20dea0b8c802d7ef,delivered,2018-07-24 20:41:37,2018-07-26 03:24:27,2018-07-26 14:31:00,2018-08-07 15:27:45,2018-08-13,-6 days +15:27:45
2,47770eb9100c2d0c44946d9cf07ec65d,41ce2a54c0b03bf3443c3d931a367089,delivered,2018-08-08 08:38:49,2018-08-08 08:55:23,2018-08-08 13:50:00,2018-08-17 18:06:29,2018-09-04,-18 days +18:06:29
3,949d5b44dbf5de918fe9c16f97b45f8a,f88197465ea7920adcdbec7375364d82,delivered,2017-11-18 19:28:06,2017-11-18 19:45:59,2017-11-22 13:39:59,2017-12-02 00:28:42,2017-12-15,-13 days +00:28:42
4,ad21c59c0840e6cb83a9ceb5573f8159,8ab97904e6daea8866dbdbc4fb7aad2c,delivered,2018-02-13 21:18:39,2018-02-13 22:20:29,2018-02-14 19:46:34,2018-02-16 18:17:02,2018-02-26,-10 days +18:17:02


In [11]:
order_dataset["Delay"] = order_dataset["Delay"].dt.total_seconds() / 86400
delayed_order_dataset = order_dataset[order_dataset["Delay"] > 1]
delayed_order_dataset.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,Delay
20,203096f03d82e0dffbc41ebc2e2bcfb7,d2b091571da224a1b36412c18bc3bbfe,delivered,2017-09-18 14:31:30,2017-09-19 04:04:09,2017-10-06 17:50:03,2017-10-09 22:23:46,2017-09-28,11.933171
25,fbf9ac61453ac646ce8ad9783d7d0af6,3a874b4d4c4b6543206ff5d89287f0c3,delivered,2018-02-20 23:46:53,2018-02-22 02:30:46,2018-02-26 22:25:22,2018-03-21 22:03:54,2018-03-12,9.919375
41,6ea2f835b4556291ffdc53fa0b3b95e8,c7340080e394356141681bd4c9b8fe31,delivered,2017-11-24 21:27:48,2017-11-25 00:21:09,2017-12-13 21:14:05,2017-12-28 18:59:23,2017-12-21,7.791238
57,66e4624ae69e7dc89bd50222b59f581f,684fa6da5134b9e4dab731e00011712d,delivered,2018-03-09 14:50:15,2018-03-09 15:40:39,2018-03-15 00:31:19,2018-04-03 13:28:46,2018-04-02,1.561644
58,a685d016c8a26f71a0bb67821070e398,911e4c37f5cafe1604fe6767034bf1ae,delivered,2017-03-13 18:14:36,2017-03-13 18:14:36,2017-03-22 14:03:09,2017-04-06 13:37:16,2017-03-30,7.567546


In [12]:
delayed_order_dataset.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6535 entries, 20 to 99432
Data columns (total 9 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       6535 non-null   object        
 1   customer_id                    6535 non-null   object        
 2   order_status                   6535 non-null   object        
 3   order_purchase_timestamp       6535 non-null   object        
 4   order_approved_at              6535 non-null   object        
 5   order_delivered_carrier_date   6534 non-null   object        
 6   order_delivered_customer_date  6535 non-null   datetime64[ns]
 7   order_estimated_delivery_date  6535 non-null   datetime64[ns]
 8   Delay                          6535 non-null   float64       
dtypes: datetime64[ns](2), float64(1), object(6)
memory usage: 510.5+ KB


In [13]:
order_item.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   order_id             112650 non-null  object 
 1   order_item_id        112650 non-null  int64  
 2   product_id           112650 non-null  object 
 3   seller_id            112650 non-null  object 
 4   shipping_limit_date  112650 non-null  object 
 5   price                112650 non-null  float64
 6   freight_value        112650 non-null  float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [14]:
order = delayed_order_dataset.merge(order_item,on="order_id", how="left")
order["Delay"] = np.ceil(order["Delay"]) 
order.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,Delay,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
0,203096f03d82e0dffbc41ebc2e2bcfb7,d2b091571da224a1b36412c18bc3bbfe,delivered,2017-09-18 14:31:30,2017-09-19 04:04:09,2017-10-06 17:50:03,2017-10-09 22:23:46,2017-09-28,12.0,1,5ac9d9e379c606e36a8094a6046f75dc,633ecdf879b94b5337cca303328e4a25,2017-09-25 04:04:09,109.90,8.96
1,fbf9ac61453ac646ce8ad9783d7d0af6,3a874b4d4c4b6543206ff5d89287f0c3,delivered,2018-02-20 23:46:53,2018-02-22 02:30:46,2018-02-26 22:25:22,2018-03-21 22:03:54,2018-03-12,10.0,1,7b717060aa783eb7f23a747a3a733dd7,c0563dd588b775f2e37747ef6ad6c92c,2018-02-28 02:30:44,109.90,15.53
2,6ea2f835b4556291ffdc53fa0b3b95e8,c7340080e394356141681bd4c9b8fe31,delivered,2017-11-24 21:27:48,2017-11-25 00:21:09,2017-12-13 21:14:05,2017-12-28 18:59:23,2017-12-21,8.0,1,be021417a6acb56b9b50d3fd2714baa8,f5f46307a4d15880ca14fab4ad9dfc9b,2017-11-30 00:21:09,339.00,17.12
3,66e4624ae69e7dc89bd50222b59f581f,684fa6da5134b9e4dab731e00011712d,delivered,2018-03-09 14:50:15,2018-03-09 15:40:39,2018-03-15 00:31:19,2018-04-03 13:28:46,2018-04-02,2.0,1,b37b72d5a56f887725c2862184b8cab8,db4350fd57ae30082dec7acbaacc17f9,2018-03-15 15:30:45,22.99,22.85
4,66e4624ae69e7dc89bd50222b59f581f,684fa6da5134b9e4dab731e00011712d,delivered,2018-03-09 14:50:15,2018-03-09 15:40:39,2018-03-15 00:31:19,2018-04-03 13:28:46,2018-04-02,2.0,2,b37b72d5a56f887725c2862184b8cab8,db4350fd57ae30082dec7acbaacc17f9,2018-03-15 15:30:45,22.99,22.85


In [15]:
order.isna().sum()

order_id                         0
customer_id                      0
order_status                     0
order_purchase_timestamp         0
order_approved_at                0
order_delivered_carrier_date     1
order_delivered_customer_date    0
order_estimated_delivery_date    0
Delay                            0
order_item_id                    0
product_id                       0
seller_id                        0
shipping_limit_date              0
price                            0
freight_value                    0
dtype: int64

In [16]:
order.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7265 entries, 0 to 7264
Data columns (total 15 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       7265 non-null   object        
 1   customer_id                    7265 non-null   object        
 2   order_status                   7265 non-null   object        
 3   order_purchase_timestamp       7265 non-null   object        
 4   order_approved_at              7265 non-null   object        
 5   order_delivered_carrier_date   7264 non-null   object        
 6   order_delivered_customer_date  7265 non-null   datetime64[ns]
 7   order_estimated_delivery_date  7265 non-null   datetime64[ns]
 8   Delay                          7265 non-null   float64       
 9   order_item_id                  7265 non-null   int64         
 10  product_id                     7265 non-null   object        
 11  seller_id        

In [17]:
len(order[order["Delay"] > 3]["seller_id"])

5724

In [18]:
seller_delay = order.groupby("seller_id")["Delay"].agg(["count","mean","max"])
seller_delay = seller_delay.sort_values("count", ascending=False)
seller_delay.head(20)

,count,mean,max
seller_id,,,
4a3ca9315b744ce9f8e9374361493884,189,12.047619,138.0
1f50f920176fa81dab994f9023523100,150,11.466667,144.0
4869f7a5dfa277a7dca6462dcf3b52b2,121,10.239669,31.0
7c67e1448b00f6e969d365cea6b010ab,120,11.900000,154.0
6560211a19b47992c3666cc44a7e94c0,102,8.549020,66.0
ea8482cd71df3c1969d7b9473ff13abc,100,9.100000,110.0
da8622b14eb17ae2831f4ac5b9dab84a,99,11.848485,97.0
955fee9216a65b617aa5c0531780ce60,95,8.915789,53.0
1025f0e2d44d7041d6cf58b6550e0bfa,95,8.063158,75.0


In [19]:
order.groupby(pd.cut(order["Delay"], bins=[0,3,7,15,999]))["freight_value"].mean()

C:\Users\RITHIK KUMAR\AppData\Local\Temp\ipykernel_10192\681504762.py:1: FutureWarning: The default of observed=False is deprecated and will be changed to True in a future version of pandas. Pass observed=False to retain current behavior or observed=True to adopt the future default and silence this warning.
  order.groupby(pd.cut(order["Delay"], bins=[0,3,7,15,999]))["freight_value"].mean()


Delay
(0, 3]       20.755724
(3, 7]       22.023490
(7, 15]      23.690801
(15, 999]    24.190242
Name: freight_value, dtype: float64

In [20]:
order["Delay"].describe()
order["Delay"].quantile([0.25, 0.5, 0.75, 0.90, 0.95])

0.25     4.0
0.50     8.0
0.75    14.0
0.90    23.0
0.95    32.0
Name: Delay, dtype: float64

In [21]:
top20_count = seller_delay.head(20)["count"].sum()
print(f"{round(top20_count / len(order) * 100, 2)}%")

24.64%


In [22]:
order["seller_id"].nunique()

1274

In [23]:
1274*0.24

305.76

In [24]:
top20_sellers = seller_delay.head(20).index.tolist()
top20_revenue = order[order["seller_id"].isin(top20_sellers)]["price"].sum()
total_revenue = order["price"].sum()
print(round(top20_revenue / total_revenue * 100, 2))

19.31


In [25]:
item_and_payment = order_item.merge(order_payment, on ="order_id", how = "left")
item_and_payment.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,payment_sequential,payment_type,payment_installments,payment_value
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29,1.0,credit_card,2.0,72.19
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93,1.0,credit_card,3.0,259.83
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87,1.0,credit_card,5.0,216.87
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79,1.0,credit_card,2.0,25.78
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14,1.0,credit_card,3.0,218.04


In [26]:
item_and_payment.isna().sum()

order_id                0
order_item_id           0
product_id              0
seller_id               0
shipping_limit_date     0
price                   0
freight_value           0
payment_sequential      3
payment_type            3
payment_installments    3
payment_value           3
dtype: int64

In [27]:
item_and_payment["thirty_%_of_price"] = item_and_payment["price"] * 0.30
item_and_payment.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,payment_sequential,payment_type,payment_installments,payment_value,thirty_%_of_price
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29,1.0,credit_card,2.0,72.19,17.670
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93,1.0,credit_card,3.0,259.83,71.970
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87,1.0,credit_card,5.0,216.87,59.700
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79,1.0,credit_card,2.0,25.78,3.897
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14,1.0,credit_card,3.0,218.04,59.970


In [28]:
item_and_payment[item_and_payment["freight_value"] >= item_and_payment["thirty_%_of_price"]]

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,payment_sequential,payment_type,payment_installments,payment_value,thirty_%_of_price
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79,1.0,credit_card,2.0,25.78,3.897
5,00048cc3ae777c65dbb7d2a0634bc1ea,1,ef92defde845ab8450f9d70c526ef70f,6426d21aca402a131fc0a5d0960a3c90,2017-05-23 03:55:27,21.90,12.69,1.0,boleto,1.0,34.59,6.570
6,00054e8431b9d7675808bcb819fb4a32,1,8d4f2bb7e93e6710a28f34fa83ee7d28,7040e82f899a04d1b434b795a43b4617,2017-12-14 12:10:31,19.90,11.85,1.0,credit_card,1.0,31.75,5.970
12,0006ec9db01a64e59a68b2c340bf65a7,1,99a4788cb24856965c36a24e339b6058,4a3ca9315b744ce9f8e9374361493884,2018-07-26 17:24:20,74.00,23.32,1.0,credit_card,4.0,97.32,22.200
19,000e562887b1f2006d75e0be9558292e,1,5ed9eaf534f6936b51d0b6c5e4d5c2e9,8cbac7e12637ed9cffa18c7875207478,2018-02-28 12:08:37,25.00,16.11,1.0,credit_card,4.0,41.11,7.500
...,...,...,...,...,...,...,...,...,...,...,...,...
117593,fffb2ef8874127f75b52b643880fd7e0,1,ebceb9726cf5c6187b48e17fe58747b0,609e1a9a6c2539919b8205cf7c4e6ff0,2018-04-09 05:08:33,17.90,22.06,1.0,boleto,1.0,39.96,5.370
117594,fffb9224b6fc7c43ebb0904318b10b5f,1,43423cdffde7fda63d0414ed38c11a73,b1fc4f64df5a0e8b6913ab38803c57a9,2017-11-03 02:55:58,55.00,34.19,1.0,boleto,1.0,356.76,16.500
117595,fffb9224b6fc7c43ebb0904318b10b5f,2,43423cdffde7fda63d0414ed38c11a73,b1fc4f64df5a0e8b6913ab38803c57a9,2017-11-03 02:55:58,55.00,34.19,1.0,boleto,1.0,356.76,16.500
117596,fffb9224b6fc7c43ebb0904318b10b5f,3,43423cdffde7fda63d0414ed38c11a73,b1fc4f64df5a0e8b6913ab38803c57a9,2017-11-03 02:55:58,55.00,34.19,1.0,boleto,1.0,356.76,16.500


In [29]:
(len(item_and_payment[item_and_payment["freight_value"] >= item_and_payment["thirty_%_of_price"]])/len(item_and_payment))*100

36.62630522771334

In [30]:
print(len(item_and_payment))

117604


In [31]:
# What % have freight >= 50% of price (freight costs half the product value)
high_drain = item_and_payment[item_and_payment["freight_value"] >= item_and_payment["price"] * 0.50]
print(round(len(high_drain) / len(item_and_payment) * 100, 2))

# What % have freight >= 100% of price (freight costs MORE than the product)
extreme_drain = item_and_payment[item_and_payment["freight_value"] >= item_and_payment["price"]]
print(round(len(extreme_drain) / len(item_and_payment) * 100, 2))

16.92
3.72


In [32]:
item_and_payment["freight_ratio"] = item_and_payment["freight_value"] / item_and_payment["price"]
item_and_payment.groupby("seller_id")["freight_ratio"].mean().sort_values(ascending=False).head(10)

seller_id
3d62f86afa7c73be2628a3be1423f5a0    3.535000
2709af9587499e95e803a6498a5a56e9    3.123362
9cea4682e16655f7da87f406902e989e    2.509630
cf6f6bc4df3999b9c6440f124fb2f687    2.491429
db1f0f7c7e5e87f40b62226f020bb297    2.302388
0f94588695d71662beec8d883ffacf09    2.065556
ce616e1913288884e7742faac9d981db    1.974836
96804ea39d96eb908e7c3afdb671bb9e    1.926344
80146bc52c4f6af5239b23de2a20262b    1.833210
95cca791657aabeff15a07eb152d7841    1.824825
Name: freight_ratio, dtype: float64

In [33]:
top20_delay_sellers = set(seller_delay.head(20).index.tolist())
top10_freight_sellers = set(item_and_payment.groupby("seller_id")["freight_ratio"].mean().sort_values(ascending=False).head(10).index.tolist())

overlap = top20_delay_sellers.intersection(top10_freight_sellers)
print(f"Overlapping sellers: {len(overlap)}")
print(overlap)

Overlapping sellers: 0
set()


In [34]:
top50_freight_sellers = set(item_and_payment.groupby("seller_id")["freight_ratio"].mean().sort_values(ascending=False).head(50).index.tolist())
overlap2 = top20_delay_sellers.intersection(top50_freight_sellers)
print(len(overlap2))

0


In [35]:
order_review_seller_id  = order_item.merge(order_review, on = "order_id", how = "left")
order_review_seller_id.head()

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,review_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
0,00010242fe8c5a6d1ba2dd792cb16214,1,4244733e06e7ecb4970a6e2683c13e61,48436dade18ac8b2bce089ec2a041202,2017-09-19 09:45:35,58.90,13.29,97ca439bc427b48bc1cd7177abe71365,5.0,NaN,"Perfeito, produto entregue antes do combinado.",2017-09-21 00:00:00,2017-09-22 10:57:03
1,00018f77f2f0320c557190d7a144bdd3,1,e5f2d52b802189ee658865ca93d83a8f,dd7ddc04e1b6c2c614352b383efe2d36,2017-05-03 11:05:13,239.90,19.93,7b07bacd811c4117b742569b04ce3580,4.0,NaN,NaN,2017-05-13 00:00:00,2017-05-15 11:34:13
2,000229ec398224ef6ca0657da4fc703e,1,c777355d18b72b67abbeef9df44fd0fd,5b51032eddd242adc84c38acab88f23d,2018-01-18 14:48:30,199.00,17.87,0c5b33dea94867d1ac402749e5438e8b,5.0,NaN,Chegou antes do prazo previsto e o produto sur...,2018-01-23 00:00:00,2018-01-23 16:06:31
3,00024acbcdf0a6daa1e931b038114c75,1,7634da152a4610f1595efa32f14722fc,9d7a1d34a5052409006425275ba1c2b4,2018-08-15 10:10:18,12.99,12.79,f4028d019cb58564807486a6aaf33817,4.0,NaN,NaN,2018-08-15 00:00:00,2018-08-15 16:39:01
4,00042b26cf59d7ce69dfabb4e55b4fd9,1,ac6c3623068f30de03045865e4e10089,df560393f3a51e74553ab94004ba5c87,2017-02-13 13:57:51,199.90,18.14,940144190dcba6351888cafa43f3a3a5,5.0,NaN,Gostei pois veio no prazo determinado .,2017-03-02 00:00:00,2017-03-03 10:54:59


In [36]:
print(max(order_review_seller_id["review_score"]))
print(min(order_review_seller_id["review_score"]))

5.0
1.0


In [37]:
top_20_bad_review_sellers = set(order_review_seller_id[order_review_seller_id["review_score"] < 3].groupby("seller_id")["review_score"]
                                .count().sort_values(ascending=False).head(20).index.to_list())

In [38]:
delay_seller_bad_review_sellers = top20_delay_sellers.intersection(top_20_bad_review_sellers)
print(len(delay_seller_bad_review_sellers))
delay_seller_bad_review_sellers

15


{'1025f0e2d44d7041d6cf58b6550e0bfa',
 '1835b56ce799e6a4dc4eddc053f04066',
 '1f50f920176fa81dab994f9023523100',
 '3d871de0142ce09b7081e2b9d1733cb1',
 '4869f7a5dfa277a7dca6462dcf3b52b2',
 '4a3ca9315b744ce9f8e9374361493884',
 '620c87c171fb2a6dd6e8bb4dec959fc6',
 '6560211a19b47992c3666cc44a7e94c0',
 '7a67c85e85bb2ce8582c35f2203ad736',
 '7c67e1448b00f6e969d365cea6b010ab',
 '8b321bb669392f5163d04c59e235e066',
 '955fee9216a65b617aa5c0531780ce60',
 'cc419e0650a3c5ba77189a1882b7556a',
 'da8622b14eb17ae2831f4ac5b9dab84a',
 'ea8482cd71df3c1969d7b9473ff13abc'}

In [39]:
bad_reviews_from_delay_sellers = order_review_seller_id[
    (order_review_seller_id["review_score"] < 3) &
    (order_review_seller_id["seller_id"].isin(top20_delay_sellers))
]
print(round(len(bad_reviews_from_delay_sellers) / 18109 * 100, 2))

22.04


In [46]:
df = order.merge(customer_data, on = "customer_id", how = "left")
df.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,Delay,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state
0,203096f03d82e0dffbc41ebc2e2bcfb7,d2b091571da224a1b36412c18bc3bbfe,delivered,2017-09-18 14:31:30,2017-09-19 04:04:09,2017-10-06 17:50:03,2017-10-09 22:23:46,2017-09-28,12.0,1,5ac9d9e379c606e36a8094a6046f75dc,633ecdf879b94b5337cca303328e4a25,2017-09-25 04:04:09,109.90,8.96,d699688533772c15a061e8ce81cb56df,4001,sao paulo,SP
1,fbf9ac61453ac646ce8ad9783d7d0af6,3a874b4d4c4b6543206ff5d89287f0c3,delivered,2018-02-20 23:46:53,2018-02-22 02:30:46,2018-02-26 22:25:22,2018-03-21 22:03:54,2018-03-12,10.0,1,7b717060aa783eb7f23a747a3a733dd7,c0563dd588b775f2e37747ef6ad6c92c,2018-02-28 02:30:44,109.90,15.53,a25d5f94840d3c6a1a49f271ed83f4ec,21715,rio de janeiro,RJ
2,6ea2f835b4556291ffdc53fa0b3b95e8,c7340080e394356141681bd4c9b8fe31,delivered,2017-11-24 21:27:48,2017-11-25 00:21:09,2017-12-13 21:14:05,2017-12-28 18:59:23,2017-12-21,8.0,1,be021417a6acb56b9b50d3fd2714baa8,f5f46307a4d15880ca14fab4ad9dfc9b,2017-11-30 00:21:09,339.00,17.12,3e4fd73f1e86b135b9b121d6abbe9597,19400,presidente venceslau,SP
3,66e4624ae69e7dc89bd50222b59f581f,684fa6da5134b9e4dab731e00011712d,delivered,2018-03-09 14:50:15,2018-03-09 15:40:39,2018-03-15 00:31:19,2018-04-03 13:28:46,2018-04-02,2.0,1,b37b72d5a56f887725c2862184b8cab8,db4350fd57ae30082dec7acbaacc17f9,2018-03-15 15:30:45,22.99,22.85,ddf60e20e6e262e2136801ce5cd628b0,49030,aracaju,SE
4,66e4624ae69e7dc89bd50222b59f581f,684fa6da5134b9e4dab731e00011712d,delivered,2018-03-09 14:50:15,2018-03-09 15:40:39,2018-03-15 00:31:19,2018-04-03 13:28:46,2018-04-02,2.0,2,b37b72d5a56f887725c2862184b8cab8,db4350fd57ae30082dec7acbaacc17f9,2018-03-15 15:30:45,22.99,22.85,ddf60e20e6e262e2136801ce5cd628b0,49030,aracaju,SE


In [48]:
df = df.merge(seller_details, on = "seller_id", how = "left")
df.head()

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date,Delay,order_item_id,...,shipping_limit_date,price,freight_value,customer_unique_id,customer_zip_code_prefix,customer_city,customer_state,seller_zip_code_prefix,seller_city,seller_state
0,203096f03d82e0dffbc41ebc2e2bcfb7,d2b091571da224a1b36412c18bc3bbfe,delivered,2017-09-18 14:31:30,2017-09-19 04:04:09,2017-10-06 17:50:03,2017-10-09 22:23:46,2017-09-28,12.0,1,...,2017-09-25 04:04:09,109.90,8.96,d699688533772c15a061e8ce81cb56df,4001,sao paulo,SP,4438,sao paulo,SP
1,fbf9ac61453ac646ce8ad9783d7d0af6,3a874b4d4c4b6543206ff5d89287f0c3,delivered,2018-02-20 23:46:53,2018-02-22 02:30:46,2018-02-26 22:25:22,2018-03-21 22:03:54,2018-03-12,10.0,1,...,2018-02-28 02:30:44,109.90,15.53,a25d5f94840d3c6a1a49f271ed83f4ec,21715,rio de janeiro,RJ,9220,santo andre,SP
2,6ea2f835b4556291ffdc53fa0b3b95e8,c7340080e394356141681bd4c9b8fe31,delivered,2017-11-24 21:27:48,2017-11-25 00:21:09,2017-12-13 21:14:05,2017-12-28 18:59:23,2017-12-21,8.0,1,...,2017-11-30 00:21:09,339.00,17.12,3e4fd73f1e86b135b9b121d6abbe9597,19400,presidente venceslau,SP,89165,rio do sul,SC
3,66e4624ae69e7dc89bd50222b59f581f,684fa6da5134b9e4dab731e00011712d,delivered,2018-03-09 14:50:15,2018-03-09 15:40:39,2018-03-15 00:31:19,2018-04-03 13:28:46,2018-04-02,2.0,1,...,2018-03-15 15:30:45,22.99,22.85,ddf60e20e6e262e2136801ce5cd628b0,49030,aracaju,SE,3126,sao paulo,SP
4,66e4624ae69e7dc89bd50222b59f581f,684fa6da5134b9e4dab731e00011712d,delivered,2018-03-09 14:50:15,2018-03-09 15:40:39,2018-03-15 00:31:19,2018-04-03 13:28:46,2018-04-02,2.0,2,...,2018-03-15 15:30:45,22.99,22.85,ddf60e20e6e262e2136801ce5cd628b0,49030,aracaju,SE,3126,sao paulo,SP


In [49]:
df.columns

Index(['order_id', 'customer_id', 'order_status', 'order_purchase_timestamp',
       'order_approved_at', 'order_delivered_carrier_date',
       'order_delivered_customer_date', 'order_estimated_delivery_date',
       'Delay', 'order_item_id', 'product_id', 'seller_id',
       'shipping_limit_date', 'price', 'freight_value', 'customer_unique_id',
       'customer_zip_code_prefix', 'customer_city', 'customer_state',
       'seller_zip_code_prefix', 'seller_city', 'seller_state'],
      dtype='object')

In [57]:
state_delay = df.groupby("customer_state")["Delay"].agg(["mean", "count"]).reset_index()
state_delay.columns = ["customer_state","Delay_avg", "Delay_cnt"]
print(min(state_delay["Delay_cnt"]))
print(state_delay["Delay_cnt"].median())
state_delay.sort_values(by = "Delay_cnt",ascending=False).head(20)

3
119.0


,customer_state,Delay_avg,Delay_cnt
25,SP,9.183554,2043
18,RJ,14.255474,1644
10,MG,9.416961,566
4,BA,12.792237,438
22,RS,11.756983,358
23,SC,9.862687,335
7,ES,11.799163,239
17,PR,9.095890,219
5,CE,15.695876,194
15,PE,12.729560,159


In [60]:
cross_state = df.groupby(["seller_state", "customer_state"])["Delay"].agg(["mean","count"]).reset_index()
cross_state.columns = ["seller_state", "customer_state", "avg_delay", "order_count"]
cross_state = cross_state[cross_state["order_count"] >= 30]
cross_state = cross_state.sort_values("avg_delay", ascending=False)
cross_state.head(15)

,seller_state,customer_state,avg_delay,order_count
69,MG,BA,21.631579,38
179,SC,RJ,16.265306,49
119,PR,RJ,15.766129,124
204,SP,RN,15.194444,36
209,SP,SE,14.906977,43
201,SP,PI,14.644068,59
190,SP,CE,14.234899,149
203,SP,RJ,14.150512,1269
83,MG,RJ,13.872549,102
112,PR,MG,13.710526,38


# Olist Marketplace — Supply Chain Analytics Findings

## Finding 1: Delay is a Seller Concentration Problem
A disproportionate share of delivery failures originates from a critically small seller segment. Just 20 sellers — representing 1.57% of the delayed seller base — are responsible for 24.6% of all delayed orders on the platform. These sellers are not peripheral contributors; they collectively generate 19.3% of total marketplace revenue, creating a direct financial dependency that prevents simple removal. This is not a platform-wide operational failure — it is a concentrated seller performance problem that the business has structurally allowed to persist because of its revenue exposure to these same sellers. The recommended intervention is a 90-day performance remediation program targeting these 20 sellers specifically, with defined delay thresholds and contract exit criteria as consequences for non-compliance.

## Finding 2: Freight Cost Structure is Silently Draining Margins
The marketplace has a systemic freight-to-price ratio problem that is not visible in top-line revenue but is actively eroding order-level profitability. 36.6% of all orders carry freight costs exceeding 30% of the product price — a margin-stressed threshold that makes a large portion of the catalog economically fragile. More critically, 3.72% of orders are being fulfilled at a net loss, where freight costs exceed the product price entirely. The worst freight-ratio sellers show average freight costs 2.5 to 3.5 times the product value. This is a pricing structure failure, not a logistics efficiency problem. The fix requires implementing minimum order value policies and freight-to-price ratio caps by product category to eliminate loss-making order fulfillment.

## Finding 3: Delay and Freight Failures are Structurally Independent — and Compounded by Customer Damage
Cross-referencing the top 20 delay sellers against the top 50 highest freight-ratio sellers reveals zero overlap — these are two entirely separate seller populations driving two entirely separate failure modes simultaneously. The business cannot solve both problems with a single intervention. However, the delay problem carries a customer experience dimension that amplifies its damage beyond operations: 75% of the top 20 delay sellers (15 out of 20) also rank among the worst-reviewed sellers on the platform, and these 20 sellers alone account for 22% of all bad reviews (score < 3) across the entire marketplace. Operational failure is visibly translating into customer dissatisfaction at scale, meaning the revenue dependency on these sellers comes with a compounding hidden cost in customer trust erosion.

## Finding 4: Geographic Delay Patterns Reveal a Logistics Infrastructure Gap
State-level and route-level analysis separates seller-driven delays from infrastructure-driven delays. Rio de Janeiro receives delayed shipments from four different origin states — SC, PR, SP, and MG — with average delays of 14 to 16 days regardless of which seller fulfilled the order. This rules out seller performance as the cause and points directly to last-mile carrier failures into RJ. Similarly, SP→Northeast corridors (CE, SE, PI, RN) consistently average 14 to 15 days of delay across high order volumes, indicating that northeastern Brazil is structurally underserved by the current carrier network. The single worst route, MG→BA, averages 21.6 days. These are not seller problems — they require logistics partner renegotiation, regional carrier expansion, or fulfillment center investment in underserved corridors.

- The bad review rate for these 20 sellers stands at 17.29% versus 15.65% for the rest of the platform — confirming that their operational failures translate into measurably worse customer experience, not just slower deliveries.

In [61]:
# bad review rate for top 20 delay sellers vs rest of platform
top20_reviews = order_review_seller_id[order_review_seller_id["seller_id"].isin(top20_delay_sellers)]
others_reviews = order_review_seller_id[~order_review_seller_id["seller_id"].isin(top20_delay_sellers)]

top20_bad_rate = (top20_reviews["review_score"] < 3).sum() / len(top20_reviews)
others_bad_rate = (others_reviews["review_score"] < 3).sum() / len(others_reviews)

print(round(top20_bad_rate * 100, 2))
print(round(others_bad_rate * 100, 2))

17.29
15.65
